### ISSUE TITLE
Implement reversible JSON serialization of Numpy arrays

### ISSUE THREAD

**Issue by kevinsung** [quantumlib/Cirq#2742]
Currently, it is not the case that
```
cirq.read_json(json_text=cirq.to_json(x))
```
always returns something equivalent to `x`, or even has the same type as `x`. It is not the case when `x` is a Numpy array, since the line above would return a list. But reversibility certainly seems to me like something that should be required of a serialization protocol. @mpharrigan @Strilanc

**Comment by mpharrigan** [comment]
eh I always envisioned JSON as a best-faith attempt to provide an accurate and interoperable representation of cirq objects.

Relatedly:
 - no distinction between tuples and lists
 - dictionaries need string keys or you need to manually send through `list(x.items())` and `dict(x)`
 - numbers don't keep track of whether they were floats or ints

If you want 1:1 serialization you are always free to use `pickle`. I think this clearly highlights the tradeoff between general/interoperable/lossy representation of python objects with specific/lossless representation.

**Comment by dabacon** [comment]
Is there something concrete in Cirq we could do here?  Does the issue with numpy arrays come up in our objects?

**Comment by mpharrigan** [comment]
I think the only action item would be to document the above limitations. If any of our objects don't handle numpy nuances, that would be a bug and we should fix it; but I don't think there are any.

**Comment by kevinsung** [comment]
I'm actually in favor of implementing reversible serialization of Numpy arrays; this would be extremely useful, and the serialization protocol of TrialResult already does it.

**Comment by mpharrigan** [comment]
How do you want to do that?

**Comment by kevinsung** [comment]
The implementation would be the same as that in TrialResult to serialize the measurements array. It is very similar to the implementation in https://github.com/quantumlib/Cirq/pull/2712. But instead of wrapping arrays in a separate object, as proposed in https://github.com/quantumlib/Cirq/pull/2712, we just serialize the arrays directly.

**Comment by mpharrigan** [comment]
i.e. modify the `cirq.to_json` `CirqEncoder`  to check for `np.ndarray`?

**Comment by kevinsung** [comment]
Yup. That is, treat Numpy arrays as we do other non-Cirq objects that we serialize, like Pandas dataframes, Sympy symbols, etc.

**Comment by mpharrigan** [comment]
My original motivation was: typically when running an experiment you have a set of parameters that you may be sweeping over or similar. You've generated the set with `np.linspace` so its a numpy array. You save it with your results and later when you're trying to remember what set of parameters you ran you can pull up the json and quickly scan for the values instead of trying to decode a hexdump.

Said another way: I posited that most of the time, numpy arrays are relatively small and list-like. In the other case when you have a big ol' blob of binary data, you have to explicitly encode it (or have a big json file).

That was the original logic. If that's not the case, we can switch it

**Comment by kevinsung** [comment]
For most of my use cases, the Numpy arrays contain a large amount of numerical data (such as state vector amplitudes), so space efficiency and reversibility are crucial. I think that for cases when the data should be human-readable, it's okay to force the user to convert to a list or other type.

**Comment by balopat** [comment]
Discussed on Cirq Cynq:

- it doesn't feel like that it's cirq's responsibility to provide this reversible serialization for naked numpy arrays - it is however when it's embedded in a Cirq object like `cirq.Result` - for which we have helper functions
- changing the default here would be a breaking change
- We could expose these helper functions to be more discoverable though. Marking as triage/accepted for this scope.

**Comment by dstrain115** [comment]
Support for this has improved since 2020.  Currently, I believe, numpy arrays are serialized into a string form if they are found as attributes of cirq objects.  Closing this as obsolete.  Please re-open if you find issues.
